lib_free_u/global_state.py

```diff
@@ -63,12 +66,18 @@ def group_stage_infos(self):
 
         return res
 
+    def to_dict(self):
+        result = vars(self).copy()
+        result["stage_infos"] = [stage_info.to_dict() for stage_info in result["stage_infos"]]
+        del result["enable"]
+        return result
+
     def update(self, state):
         for k, v in vars(state).items():
             self.update_attr(k, v)
 
     def update_attr(self, key, value):
-        if key in xyz_locked_attrs:
+        if xyz_locked or key in xyz_locked_attrs:
```

Comment:
is there a reason we prevent updating any attribute? `process` needs to pass through for all attributes not currently active in XYZ grid.

Also, you can use `bool(xyz_locked_attrs)` if you need to check whether any attribute of the script is currently active in xyz grid. Defining `global_state.xyz_locked` makes the code not orthogonal IIUC.

Reply:
So without locking attributes, the XYZ plot wasn't working. Because code uses silly single characters for sending properties to xyz it is hard to bulk lock every field.

So mirroring single property locking when single property is set for XYZ, I'm locking whole structure for XYZ plot. This get self unlocked after each run (just like properties), so there shouldn't be much of a difference... I'm open for changes 😄

Reply:
"process needs to pass through for all attributes not currently active in XYZ grid." - are there any like that if whole preset is used?

Reply:
I understand now, when automating presets in XYZ grid we do need to lock everything.

> are there any [pass through attributes] if whole preset is used?
Except for the "enable" property, I think not. "enable" is not a part of preset state, so IIUC we need to account for that one.

Maybe the property should be named `xyz_preset_locked` to tie the intent to presets?

Reply:
Also, what if we xyz grid another property at the same time as presets? When using concurrent xyz grid attributes, the property should take priority over the corresponding value in the xyz preset.

Reply:
I'm not set on a solution or the internals as much, as long as the feature is working well I'm ok with changing it to match your standards. Your internal framework is great and easy to extend, so don't want to mess it too much beside to showcase new possibilities 😃